### PR TITLE
Update get_applicable_matrix_reduction() to combine dimensions of 1 with the given reduction axes.

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Update get_applicable_matrix_reduction() to combine dimensions of 1 with the given reduction axes.

**Motivation and Context**
Performance improvement for ReduceSum.
Handle more cases with the optimized matrix reduction functions.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc [hunk 1]
```diff
@@ -13,45 +15,86 @@
...
+ApplicableMatrixReduction get_applicable_matrix_reduction(
+    const cudnnReduceTensorOp_t cudnn_reduce_op,
+    const std::vector<int64_t>& dims, const std::vector<int64_t>& original_axes,
+    int& m_out, int& n_out) {
+  if (cudnn_reduce_op != CUDNN_REDUCE_TENSOR_ADD) {
+    return ApplicableMatrixReduction::None;
+  }
+
+  const auto rank = gsl::narrow<int64_t>(dims.size());
+  const auto minmax_axes = GetMinMaxContiguousAxes(rank, dims, original_axes);
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:97]: >minmax_axes [](start = 13, length = 11)

`minmax -> min_max`. minmax has a special meaning in optimization.

**Comment by wschin** [comment]
>   const int64_t m_end_axis = axes_from_beginning ? max_axis + 1 : min_axis;

m_end_axis -> row_end_axis. Also add a comment `// The axis index right after the last flattened into matrix rows.`

---
Refers to: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:115 in 3d63389. [](commit_id = 3d63389f4ab5f102e2929904683828f50f7310e9, deletion_comment = False)

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc [hunk 2]
```diff
@@ -13,45 +15,86 @@
...
-      }
+    // if all dimension values between adjacent axes are 1,
+    // treat the axes as contiguous
+    if (std::all_of(dims.begin() + *a + 1, dims.begin() + *b, is_dim_one)) {
+      continue;
     }
 
-    return std::make_pair(axes.front(), axes.back());
+    // otherwise, not contiguous
+    return nullopt;
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:60]: >nullopt [](start = 11, length = 7)

Not sure what's the convention here. Would it be `std::nullopt` or just `nullopt`.

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc [hunk 3]
```diff
@@ -13,45 +15,86 @@
...
+  const std::vector<int64_t> axes = [&original_axes, rank]() {
+    std::vector<int64_t> result(original_axes);
+    std::for_each(
+        result.begin(), result.end(),
+        [rank](int64_t& axis) { axis = HandleNegativeAxis(axis, rank); });
+    std::sort(result.begin(), result.end());
+    return result;
+  }();
+
+  assert(!axes.empty());
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:39]: Aldo do `assert(dims.size() == rank)` for safety.

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc [hunk 4]
```diff
@@ -3,7 +3,9 @@
 
 #include "core/providers/cuda/reduction/reduction_functions.h"
 
+#include <cassert>
 #include <algorithm>
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:7]: How do we determine the order of headers? Are they sorted by alphabets?

**Review by wschin — DISMISSED** [review]
:shipit:

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc (hunk 2 above)
Comment (edgchen1) [reply onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:60]: i'd prefer std::nullopt, but we don't have C++17's std::optional so this is actually onnxruntime::nullopt

---
In reply to: [520194944](https://github.com/microsoft/onnxruntime/pull/5734#discussion_r520194944) [](ancestors = 520194944)

**Comment by edgchen1** [comment]
>   const int64_t m_end_axis = axes_from_beginning ? max_axis + 1 : min_axis;

i'd prefer the original name since we refer to "m" and "n" here already. will add the comment

---
In reply to: [724353184](https://github.com/microsoft/onnxruntime/pull/5734#issuecomment-724353184) [](ancestors = 724353184)

---
Refers to: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:115 in 3d63389. [](commit_id = 3d63389f4ab5f102e2929904683828f50f7310e9, deletion_comment = False)

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cc (hunk 3 above)
Comment (edgchen1) [reply onnxruntime/core/providers/cuda/reduction/reduction_functions.cc:51]: added above

**Review by wschin — APPROVED** [review]
:shipit:

**Pull request merged**